File: plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/offline/OfflineRegionDetailActivity.java
package com.mapbox.mapboxsdk.plugins.testapp.activity.offline;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.mapbox.mapboxsdk.camera.CameraPosition;
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory;
import com.mapbox.mapboxsdk.maps.MapView;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
import com.mapbox.mapboxsdk.offline.OfflineManager;
import com.mapbox.mapboxsdk.offline.OfflineRegion;
import com.mapbox.mapboxsdk.offline.OfflineRegionStatus;
import com.mapbox.mapboxsdk.offline.OfflineTilePyramidRegionDefinition;
import com.mapbox.mapboxsdk.plugins.offline.OfflineDownload;
import com.mapbox.mapboxsdk.plugins.offline.OfflineDownloadChangeListener;
import com.mapbox.mapboxsdk.plugins.offline.OfflinePlugin;
import com.mapbox.mapboxsdk.plugins.offline.OfflineUtils;
import com.mapbox.mapboxsdk.plugins.testapp.R;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import timber.log.Timber;

/**
 * Activity showing the detail of an offline region.
 * <p>
 * This Activity can bind to the OfflineDownloadService and show progress.
 * </p>
 * <p>
 * This Activity listens to broadcast events related to successful, canceled and errored download.
 * </p>
 */
public class OfflineRegionDetailActivity extends AppCompatActivity implements OfflineDownloadChangeListener {

  public static final String KEY_REGION_ID_BUNDLE = "com.mapbox.mapboxsdk.plugins.offline.bundle.id";

  @BindView(R.id.mapView)
  MapView mapView;

  @BindView(R.id.fab_delete)
  FloatingActionButton deleteView;

  @BindView(R.id.region_state)
  TextView stateView;

  @BindView(R.id.region_state_progress)
  ProgressBar progressBar;

  @BindView(R.id.region_name)
  TextView nameView;

  @BindView(R.id.region_style_url)
  TextView styleView;

  @BindView(R.id.region_min_zoom)
  TextView minZoomView;

  @BindView(R.id.region_max_zoom)
  TextView maxZoomView;

  @BindView(R.id.region_lat_lng_bounds)
  TextView latLngBoundsView;

  private OfflinePlugin offlinePlugin;
  private OfflineRegion offlineRegion;
  private boolean isDownloading;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_offline_region_detail);
    ButterKnife.bind(this);
    mapView.onCreate(savedInstanceState);
    offlinePlugin = OfflinePlugin.getInstance();

    Bundle bundle = getIntent().getExtras();
    if (bundle != null) {
      loadOfflineDownload(bundle);
    }
  }

  private void loadOfflineDownload(Bundle bundle) {
    long regionId;
    OfflineDownload offlineDownload = bundle.getParcelable(OfflineDownload.KEY_BUNDLE);
    if (offlineDownload != null) {
      // coming from notification
      regionId = offlineDownload.getRegionId();
    } else {
      // coming from list
      regionId = bundle.getLong(KEY_REGION_ID_BUNDLE, -1);
    }

    if (regionId != -1) {
      loadOfflineRegion(regionId);
    }
  }

  private void loadOfflineRegion(final long id) {
    OfflineManager.getInstance(this)
      .listOfflineRegions(new OfflineManager.ListOfflineRegionsCallback() {

        @Override
        public void onList(OfflineRegion[] offlineRegions) {
          for (OfflineRegion region : offlineRegions) {
            if (region.getID() == id) {
              offlineRegion = region;
              OfflineTilePyramidRegionDefinition definition =
                (OfflineTilePyramidRegionDefinition) region.getDefinition();
              setupUI(definition);
              return;
            }
          }
        }

        @Override
        public void onError(String error) {
          Timber.e(error);
        }
      });
  }

  private void updateFab() {
    FloatingActionButton fab = findViewById(R.id.fab_delete);
    if (isDownloading) {
      fab.setImageResource(R.drawable.ic_cancel);
      stateView.setText("DOWNLOADING");
    } else {
      fab.setImageResource(R.drawable.ic_delete);
      stateView.setText("DOWNLOADED");
    }
  }

  private void setupUI(final OfflineTilePyramidRegionDefinition definition) {
    // update map
    mapView.getMapAsync(new OnMapReadyCallback() {
      @Override
      public void onMapReady(MapboxMap mapboxMap) {
        // correct style
        mapboxMap.setStyle(definition.getStyleURL());

        // position map on top of offline region
        CameraPosition cameraPosition = OfflineUtils.getCameraPosition(definition);
        mapboxMap.moveCamera(CameraUpdateFactory.newCameraPosition(cameraPosition));

        // restrict camera movement
        mapboxMap.setMinZoomPreference(definition.getMinZoom());
        mapboxMap.setMaxZoomPreference(definition.getMaxZoom());
        mapboxMap.setLatLngBoundsForCameraTarget(definition.getBounds());
      }
    });

    // update textview data
    nameView.setText(OfflineUtils.convertRegionName(offlineRegion.getMetadata()));
    styleView.setText(definition.getStyleURL());
    latLngBoundsView.setText(definition.getBounds().toString());
    minZoomView.setText(String.valueOf(definition.getMinZoom()));
    maxZoomView.setText(String.valueOf(definition.getMaxZoom()));
    offlineRegion.getStatus(offlineRegionStatusCallback);
  }


  @OnClick(R.id.fab_delete)
  public void onFabClick(View view) {
    if (offlineRegion != null) {
      if (!isDownloading) {
        // delete download
        offlineRegion.delete(offlineRegionDeleteCallback);
      } else {
        // cancel download
        OfflineDownload offlineDownload = offlinePlugin.getActiveDownloadForOfflineRegion(offlineRegion);
        if (offlineDownload != null) {
          offlinePlugin.cancelDownload(this, offlineDownload);
          isDownloading = false;
        }
      }
      view.setVisibility(View.GONE);
    }
  }

  @Override
  public void onCreate(OfflineDownload offlineDownload) {
    Timber.e("OfflineDownload created %s", offlineDownload.hashCode());
  }

  @Override
  public void onSuccess(OfflineDownload offlineDownload) {
    isDownloading = false;
    progressBar.setVisibility(View.INVISIBLE);
    updateFab();
  }

  @Override
  public void onCancel(OfflineDownload offlineDownload) {
    finish(); // nothing to do in this screen, cancel = delete
  }

  @Override
  public void onError(OfflineDownload offlineDownload, String error, String message) {
    progressBar.setVisibility(View.INVISIBLE);
    stateView.setText("ERROR");
    Toast.makeText(this, error + message, Toast.LENGTH_LONG).show();
  }

  @Override
  public void onProgress(OfflineDownload offlineDownload, int progress) {
    if (offlineRegion == null) {
      return;
    }

    if (offlineDownload.getRegionId() == offlineRegion.getID()) {
      if (progressBar.getVisibility() != View.VISIBLE) {
        progressBar.setVisibility(View.VISIBLE);
      }
      isDownloading = true;
      progressBar.setProgress(progress);
    }
  }

  @Override
  protected void onStart() {
    super.onStart();
    mapView.onStart();
    offlinePlugin.addOfflineDownloadStateChangeListener(this);
  }

  @Override
  protected void onResume() {
    super.onResume();
    mapView.onResume();
  }

  @Override
  protected void onPause() {
    super.onPause();
    mapView.onPause();
  }

  @Override
  protected void onStop() {
    super.onStop();
    mapView.onStop();
    offlinePlugin.removeOfflineDownloadStateChangeListener(this);
  }

  @Override
  protected void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    mapView.onSaveInstanceState(outState);
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    mapView.onDestroy();
  }

  @Override
  public void onLowMemory() {
    super.onLowMemory();
    mapView.onLowMemory();
  }

  /**
   * Callback invoked when the states of an offline region changes.
   */
  private final OfflineRegion.OfflineRegionStatusCallback offlineRegionStatusCallback =
    new OfflineRegion.OfflineRegionStatusCallback() {
      @Override
      public void onStatus(OfflineRegionStatus status) {
        isDownloading = !status.isComplete();
        updateFab();
      }

      @Override
      public void onError(String error) {
        Toast.makeText(
          OfflineRegionDetailActivity.this,
          "Error getting offline region state: " + error,
          Toast.LENGTH_SHORT).show();
      }
    };

  private final OfflineRegion.OfflineRegionDeleteCallback offlineRegionDeleteCallback =
    new OfflineRegion.OfflineRegionDeleteCallback() {
      @Override
      public void onDelete() {
        Toast.makeText(
          OfflineRegionDetailActivity.this,
          "Region deleted.",
          Toast.LENGTH_SHORT).show();
        finish();
      }

      @Override
      public void onError(String error) {
        deleteView.setEnabled(true);
        Toast.makeText(
          OfflineRegionDetailActivity.this,
          "Error getting offline region state: " + error,
          Toast.LENGTH_SHORT).show();
      }
    };
}

File: app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/offline/OfflineRegionDetailActivity.kt
package com.mapbox.mapboxsdk.plugins.testapp.activity.offline

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.mapbox.mapboxsdk.offline.OfflineManager
import com.mapbox.mapboxsdk.offline.OfflineRegion
import com.mapbox.mapboxsdk.offline.OfflineRegionDefinition
import com.mapbox.mapboxsdk.offline.OfflineRegionStatus
import com.mapbox.mapboxsdk.plugins.offline.model.OfflineDownloadOptions
import com.mapbox.mapboxsdk.plugins.offline.offline.OfflineConstants.KEY_BUNDLE
import com.mapbox.mapboxsdk.plugins.offline.offline.OfflineDownloadChangeListener
import com.mapbox.mapboxsdk.plugins.offline.offline.OfflinePlugin
import com.mapbox.mapboxsdk.plugins.offline.utils.OfflineUtils
import com.mapbox.mapboxsdk.plugins.testapp.R
import kotlinx.android.synthetic.main.activity_offline_region_detail.*
import timber.log.Timber

/**
 * Activity showing the detail of an offline region.
 *
 *
 * This Activity can bind to the OfflineDownloadService and show progress.
 *
 *
 *
 * This Activity listens to broadcast events related to successful, canceled and errored download.
 *
 */
class OfflineRegionDetailActivity : AppCompatActivity(), OfflineDownloadChangeListener {

    private var offlinePlugin: OfflinePlugin? = null
    private var offlineRegion: OfflineRegion? = null
    private var isDownloading: Boolean = false

    /**
     * Callback invoked when the states of an offline region changes.
     */
    private val offlineRegionStatusCallback = object : OfflineRegion.OfflineRegionStatusCallback {
        override fun onStatus(status: OfflineRegionStatus) {
            isDownloading = !status.isComplete
            updateFab()
        }

        override fun onError(error: String) {
            Toast.makeText(
                    this@OfflineRegionDetailActivity,
                    "Error getting offline region state: $error",
                    Toast.LENGTH_SHORT).show()
        }
    }

    private val offlineRegionDeleteCallback = object : OfflineRegion.OfflineRegionDeleteCallback {
        override fun onDelete() {
            Toast.makeText(
                    this@OfflineRegionDetailActivity,
                    "Region deleted.",
                    Toast.LENGTH_SHORT).show()
            finish()
        }

        override fun onError(error: String) {
            fabDelete.isEnabled = true
            Toast.makeText(
                    this@OfflineRegionDetailActivity,
                    "Error getting offline region state: $error",
                    Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_offline_region_detail)
        mapView?.onCreate(savedInstanceState)
        offlinePlugin = OfflinePlugin.getInstance(this)

        val bundle = intent.extras
        if (bundle != null) {
            loadOfflineDownload(bundle)
        }

        fabDelete.setOnClickListener { onFabClick(it) }
    }

    private fun loadOfflineDownload(bundle: Bundle) {
        val regionId: Long
        val offlineDownload = bundle.getParcelable<OfflineDownloadOptions>(KEY_BUNDLE)
        regionId = if (offlineDownload != null) {
            // coming from notification
            offlineDownload.uuid()
        } else {
            // coming from list
            bundle.getLong(KEY_REGION_ID_BUNDLE, -1)
        }

        if (regionId != -1L) {
            loadOfflineRegion(regionId)
        }
    }

    private fun loadOfflineRegion(id: Long) {
        OfflineManager.getInstance(this)
                .listOfflineRegions(object : OfflineManager.ListOfflineRegionsCallback {

                    override fun onList(offlineRegions: Array<OfflineRegion>) {
                        for (region in offlineRegions) {
                            if (region.id == id) {
                                offlineRegion = region
                                val definition = region.definition as OfflineRegionDefinition
                                setupUI(definition)
                                return
                            }
                        }
                    }

                    override fun onError(error: String) {
                        Timber.e(error)
                    }
                })
    }

    private fun updateFab() {
        if (isDownloading) {
            fabDelete.setImageResource(R.drawable.ic_cancel)
            regionState.text = "DOWNLOADING"
        } else {
            fabDelete.setImageResource(R.drawable.ic_delete)
            regionState.text = "DOWNLOADED"
        }
    }

    private fun setupUI(definition: OfflineRegionDefinition) {
        // update map
        mapView?.getMapAsync { mapboxMap ->
            // correct style
            mapboxMap.setOfflineRegionDefinition(definition) { _ ->
                // restrict camera movement
                mapboxMap.setLatLngBoundsForCameraTarget(definition.bounds)

                // update textview data
                offlineRegion?.metadata?.let {
                    regionName.text = OfflineUtils.convertRegionName(it)
                }
                regionStyleUrl.text = definition.styleURL
                regionLatLngBounds.text = definition.bounds.toString()
                regionMinZoom.text = definition.minZoom.toString()
                regionMaxZoom.text = definition.maxZoom.toString()
                offlineRegion?.getStatus(offlineRegionStatusCallback)
            }
        }
    }

    fun onFabClick(view: View) {
        if (offlineRegion != null) {
            if (!isDownloading) {
                // delete download
                offlineRegion?.delete(offlineRegionDeleteCallback)
            } else {
                // cancel download
                val offlineDownload = offlinePlugin?.getActiveDownloadForOfflineRegion(offlineRegion)
                if (offlineDownload != null) {
                    offlinePlugin?.cancelDownload(offlineDownload)
                    isDownloading = false
                }
            }
            view.visibility = View.GONE
        }
    }

    override fun onCreate(offlineDownload: OfflineDownloadOptions) {
        Timber.e("OfflineDownloadOptions created %s", offlineDownload.hashCode())
    }

    override fun onSuccess(offlineDownload: OfflineDownloadOptions) {
        isDownloading = false
        regionStateProgress.visibility = View.INVISIBLE
        updateFab()
    }

    override fun onCancel(offlineDownload: OfflineDownloadOptions) {
        finish() // nothing to do in this screen, cancel = delete
    }

    override fun onError(offlineDownload: OfflineDownloadOptions, error: String, message: String) {
        regionStateProgress.visibility = View.INVISIBLE
        regionState.text = "ERROR"
        Toast.makeText(this, error + message, Toast.LENGTH_LONG).show()
    }

    override fun onProgress(offlineDownload: OfflineDownloadOptions, progress: Int) {
        if (offlineRegion == null) {
            return
        }

        if (offlineDownload.uuid() == offlineRegion?.id) {
            if (regionStateProgress.visibility != View.VISIBLE) {
                regionStateProgress.visibility = View.VISIBLE
            }
            isDownloading = true
            regionStateProgress.progress = progress
        }
    }

    override fun onStart() {
        super.onStart()
        mapView?.onStart()
        offlinePlugin?.addOfflineDownloadStateChangeListener(this)
    }

    override fun onResume() {
        super.onResume()
        mapView?.onResume()
    }

    override fun onPause() {
        super.onPause()
        mapView?.onPause()
    }

    override fun onStop() {
        super.onStop()
        mapView?.onStop()
        offlinePlugin?.removeOfflineDownloadStateChangeListener(this)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mapView?.onSaveInstanceState(outState)
    }

    override fun onDestroy() {
        super.onDestroy()
        mapView.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapView?.onLowMemory()
    }

    companion object {

        val KEY_REGION_ID_BUNDLE = "com.mapbox.mapboxsdk.plugins.offline.bundle.id"
    }
}